tests/functional/base.py

```diff
@@ -0,0 +1,316 @@
+import os
+import configparser
+import tempfile
+import json
+from tempest.lib.cli import base
+from tempest.lib.cli import output_parser
+from tempest.lib.common.utils import data_utils
+
+class EsiFunctionalBase(base.ClientTestBase):
+    @classmethod
+    def setUpClass(cls):
+        super(EsiFunctionalBase, cls).setUpClass()
+        cls.clients = {}
+        cls.config = {}
+        cls.client_info = {}
+        cls._cleanups = []
+
+        cls._init_config(cls)
+        cls._init_client(cls, 'admin')
+        cls._init_roles(cls)
+
+    @classmethod
+    def tearDownClass(cls):
+        cls._cleanups.reverse()
+        for cleanup in cls._cleanups:
+            cleanup[0](*cleanup[1:])
+
+    def _get_clients(self):
+        #NOTE: ClientTestBase's constructor requires this to be implemented, but
+        #to initialize our clients, we need ClientTestBase's constructor to have
+        #been called. We return nothing and leave the job up to _init_clients().
+        return {}
+
+    def _init_config(self):
+        venv_name = os.environ.get('OS_VENV_NAME', default='functional')
+        self.config['cli_dir'] = os.path.join(os.path.abspath('.'), 
+                '.tox/%s/bin' % venv_name)
+
+        # allow custom configuration to be passed in via env var
+        cfg_file_path = os.environ.get('OS_ESI_CFG_PATH',
+                '/etc/esi-leap/esi-leap.conf')
+        cfg_parser = configparser.ConfigParser()
+        if not cfg_parser.read(cfg_file_path):
+            self.fail('Could not open config file %s for reading' %
+                    cfg_file_path)
+            return -1
+
+        # attempts to read authentication credentials in this order:
+        # 1) [keystone] section of config file
+        # 2) [keystone_authtoken] section of config file
+        # 3) from the environment variables (e.g. OS_PASSWORD, etc)
+        def auth_get(val):
```

Comment:
As a rule, I'd avoid putting functions inside of functions. Maybe bring this outside of the function and call it `def _auth_get`?

Reply:
Now that I think about this, it doesn't even need to be a function really since it's only called one time; I'll just stick it in the for loop down below.